WKT Support

drewjj opened the pull request:
WKT support
- New Kafka topic added, topic.ProcessedMapWKT
- Optional WKT mode enabled by assigning the environment variable GEOMETRY_OUTPUT_MODE=WKT
- WKT mode still will output normal GeoJSON to topic.ProcessedMap for usage in other usdot-jpo-ode modules
- Unit tests added

Michael7371 commented:
Looks good to me!

iyourshaw commented:
The existing test classes:

iyourshaw commented on jpo-geojsonconverter/src/main/resources/schemas/processed-map.schema.json:
```diff
@@ -70,7 +70,7 @@
-            "type": "object",
+            "type": ["object", "string"],
             "properties": {
                 "type": {
```
Make sure this validation still works in case the geometry is a WKT string. (Add a test with a WKT example to the MapJsonValidatorTest class).

Might need to remove  

```
"required": [
                "type",
                "coordinates"
            ],
```
or add logic so those are only required if type is object.

drewjj replied:
I can confirm this works regardless of the child object requiring the type and coordinates. I can add a test case.

drewjj replied:
I am going to actually push back on this for now, we don't have any tests to verify the current processed-map output even passes validation. The method that current validators are being tested are not applicable since those operate as services and are an active part of the regular application. We could add the schemas as test resources but then we have duplication. Alternatively, we could move the schemas from the main resources to the test since they are not used in the application to begin with.

iyourshaw replied:
Point taken that the MapJsonValidatorTest class doesn't currently test validation of ProcessedMap at all.   Such tests could be added without moving or copying the schema, but as you say, since ProcessedMaps conformance with that schema isn't an integral concern of the geojson converter then it's not so important to have a unit test.

drewjj commented on jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/GeoJsonConverterProperties.java:
```diff
@@ -82,6 +85,13 @@ public void initialize() {
             }
         }
+
+        String geometryOutputMode = CommonUtils.getEnvironmentVariable("GEOMETRY_OUTPUT_MODE");
```
Addressed this.

iyourshaw replied:
If you wanted to customize that you could inject the spring property into the setter in the properties class, something like this pseudocode

```java
@Value("${geometry.output.mode}")
public void setGeometryOutputMode(String str) {
   if enum contains string save as enum and set WKT Flag bit
   else log error and set default
}
```

drewjj commented on jpo-geojsonconverter/src/main/java/us/dot/its/jpo/geojsonconverter/GeoJsonConverterProperties.java:
```diff
@@ -82,6 +85,13 @@ public void initialize() {
+
+        String geometryOutputMode = CommonUtils.getEnvironmentVariable("GEOMETRY_OUTPUT_MODE");
+        if (geometryOutputMode != null) {
+            if (geometryOutputMode.equals("WKT")) {
```
> The existing test classes:
> 
> ```
> us.dot.its.jpo.geojsonconverter.serialization.deserializers.JsonDeserializerTest
>    and
> us.dot.its.jpo.geojsonconverter.serialization.serializers.JsonSerializerTest
> ```
> 
> only have tests for the original case where the geometry is geojson. I'd suggest adding additional tests to both of those classes for the case where the geometry is a wkt string. Also fix the warnings about raw generic types in those tests, for example:
> 
> ![image](https://user-images.githubusercontent.com/39739503/236731444-d6b13edc-e78e-4820-bf31-1bc22253bb4e.png)

I have addressed this.

drewjj replied:
Addressed

drewjj commented:
> The existing test classes:
> 
> ```
> us.dot.its.jpo.geojsonconverter.serialization.deserializers.JsonDeserializerTest
>    and
> us.dot.its.jpo.geojsonconverter.serialization.serializers.JsonSerializerTest
> ```
> 
> only have tests for the original case where the geometry is geojson. I'd suggest adding additional tests to both of those classes for the case where the geometry is a wkt string. Also fix the warnings about raw generic types in those tests, for example:
> 
> ![image](https://user-images.githubusercontent.com/39739503/236731444-d6b13edc-e78e-4820-bf31-1bc22253bb4e.png)

Addressed this comment. I added a handful of new tests.

iyourshaw commented:
> > The existing test classes:
> > ```
> > us.dot.its.jpo.geojsonconverter.serialization.deserializers.JsonDeserializerTest
> >    and
> > us.dot.its.jpo.geojsonconverter.serialization.serializers.JsonSerializerTest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > only have tests for the original case where the geometry is geojson. I'd suggest adding additional tests to both of those classes for the case where the geometry is a wkt string. Also fix the warnings about raw generic types in those tests, for example:
> > ![image](https://user-images.githubusercontent.com/39739503/236731444-d6b13edc-e78e-4820-bf31-1bc22253bb4e.png)
> 
> Addressed this comment. I added a handful of new tests.

Looks good
